src/librustdoc/html/format.rs

```diff
@@ -37,114 +37,36 @@ use crate::html::render::Context;
 use crate::joined::Joined as _;
 use crate::passes::collect_intra_doc_links::UrlFragment;
 
-pub(crate) trait Print {
-    fn print(self, buffer: &mut Buffer);
-}
-
-impl<F> Print for F
-where
-    F: FnOnce(&mut Buffer),
-{
-    fn print(self, buffer: &mut Buffer) {
-        (self)(buffer)
-    }
-}
-
-impl Print for String {
-    fn print(self, buffer: &mut Buffer) {
-        buffer.write_str(&self);
-    }
-}
-
-impl Print for &'_ str {
-    fn print(self, buffer: &mut Buffer) {
-        buffer.write_str(self);
-    }
-}
+/// This macro is the same as [`std::write!`] for [`String`]s, but swallows the returned `Result`,
+/// since writing into a `String` can never fail.
+macro_rules! write_str {
```

Comment:
Please avoid adding macros to do such little operations. You can instead write a function expecting a `String` so that it doesn't need a new trait for each `write_str!`/`write!` call. Also, renaming the macro into `write` makes the code reading a lot more complex.

Reply:
The reason why this is a macro (and why `write!` itself is a macro in the STD), is that otherwise every call would like something like this:

```rust
let mut buf = String::new();
write_str(&mut buf, format_args!("format str, {}", var));
// instead of:
write!(buf, "format str, {}", var);
```

I don't like this macro either, but I think the best alternative would be to add `let _ = ...` or `unwrap()`s to every call to `write!` on a string (which would be very verbose), or leave the `Buffer` be.

For the record, my future plan is to move as many functions as possible to use RPIT to return an `impl fmt::Display`, instead of accepting a `&mut String` as a param, and then get rid of this macro. I tried to do that in #136748 , but that PR became very big...

Reply:
Still better than writing a new macro, using `format_args` is fine. As a first step until something better comes up is ok.

Reply:
Done.
Not sure I love the function names (`write_str` and `writeln_str`), if you have a better idea in mind I'll be happy to rename them :)